### PR TITLE
Add per-class loss weighting to fix bat/mammal imbalance

### PR DESCRIPTION
**Pull request by tphakala**
## What

Adds an optional `--class_loss_weight` flag: a per-species multiplier applied at the loss level, specified per taxonomic class, for example:

```
--class_loss_weight "mammalia=6.0"
```

The weight scales the per-species BCE before reduction in all species loss variants (assume-negative, focal, ASL, BCE). It is registered as a buffer so it follows `.to(device)` and is saved with the checkpoint. It defaults to off, so existing runs are unchanged.

## Why

The geomodel predicts birds well but scores bats (and mammals generally) near zero everywhere, even where they are common. The modelling root cause is class imbalance in a presence-only loss: a few thousand common bird species dominate the gradient and swamp the few hundred mammals.

The imbalance is structural in the data, not just in species counts. Counting recorded presence across the training set, the most common birds have 30 to 140 times more records than the European bats:

| species | recorded presence (cell-weeks) |
|---|---|
| House Sparrow | 283,000 |
| Mallard | 233,000 |
| Plecotus auritus (best-recorded bat) | 9,900 |
| Eptesicus nilssonii (worst) | 2,000 |

So even where a bat is present every night, GBIF holds far fewer records for it, and the loss barely sees it. Up-weighting the under-represented class at the loss level lets the optimiser attend to it without distorting anything else.

Important: this stays strictly at the loss level. An earlier attempt to encode the weight into the target value instead collapsed bat predictions to near zero, because it turned the binary occurrence target into a soft one the model then learned to suppress. The multiplier here only scales the loss term; targets stay binary.

## How

`--class_loss_weight "class=w,..."` is parsed into a per-species vector by looking up each species' taxonomic class, then passed to `MultiTaskLoss` and applied as `loss = loss * weight` (broadcast over the batch) before reduction.

Verified:

- New unit test (`tests/test_class_loss_weight.py`): the multiplier scales the per-species loss for assume-negative / focal / ASL, is a no-op at unit weight, and never mutates the target value.
- Controlled A/B retrain on fresh global GBIF, identical recipe except the flag. With `mammalia=6.0` (boosting 879 mammal species), every headline metric improved with no bird regression:

| metric | baseline (no weighting) | `mammalia=6.0` |
|---|---|---|
| GeoScore | 0.4965 | 0.5080 |
| bat presence (mean) | 0.193 | 0.213 |
| bird anchor (mean) | 0.773 | 0.817 |

Geographic contrast stayed near zero (the model did not start predicting bats everywhere), so the gain is real range signal, not decalibration.

This is the modelling-side counterpart to the taxonomy fix in #16. Together they address the bat gap described in #17: #16 stops the records being discarded, this lets the loss actually learn them.

AI assistance (Claude) was used to investigate the imbalance, run the A/B, and prepare this change.

Related: #17

### CODE REVIEW

**Comment by kahst** [comment]
I'm not sure if a new loss is the best way to do that. We do sort species observations into buckets and 1 observation counts just like 1,000 observations for a given week and location - so there is some balancing. For underrepresented species, we should be able to re-balance or close gaps with label propagation and gap smoothing instead of loss weighting, don't you think? At least, I would prefer a solution that recalibrates or extends label propagation rather than moving away from vanilla bce.

**Comment by tphakala** [comment]
Thanks, that's a fair push, and I think part of this is on me for how I worded the PR. Two things I described badly.

First, calling it "per-class loss weighting" made it sound like a new loss. It isn't. For the default `bce` path it's literally `nn.BCEWithLogitsLoss(weight=...)`, the standard per-class `weight` argument; the loss stays vanilla BCE. I double-checked: stdlib `BCEWithLogitsLoss(weight=w)` and `mean(w * unreduced_bce)` come out identical, and `MultiTaskLoss`'s bce criterion is exactly that object. So this isn't moving away from BCE, it's using BCE's own class-weight knob. The focal/ASL/AN branches just thread the same multiplier through, and it's off by default so existing runs are unchanged.

Second, the imbalance table was unclear. Those counts are already post-bucketing, cell-weeks of presence, not raw records. You're right that bucketing balances within a `(cell, week)`: one record counts the same as a thousand there. But the gap I'm pointing at is across cell-weeks. After the set-aggregation in `combine.py`, the House Sparrow is present in ~280k cell-weeks and the best-recorded bat in ~10k. Bucketing doesn't touch that, and I should have labelled the table that way.

On propagation and gap smoothing: I actually agree the data side is the more principled place to fix this, and both already exist (`--propagate_labels`, `--smooth_gaps`). The reason I didn't lean on them is that, as they stand, they can't reach this particular gap. I ran a quick check against the real functions:

- Env propagation only writes into sparse cells (`< min_obs`); observed cells are read-only sources. In a setup where a bat truly occurs in 195 well-sampled cells but is recorded in only 5, propagation added it to 0 of the 195, only to sparse fringe cells. So it can't raise bat presence in the well-sampled cells where bats actually live but are under-recorded.
- Gap smoothing fills a week-run only when it's bracketed by presences on both sides, so it needs at least 2 recorded weeks in the same cell. A bat recorded 0 or 1 times in a cell gets nothing (1 week gives 0 filled, 2 weeks fills the gap between them).

To make propagation close the bat gap you'd have to synthesize positives into observed cells from weak anchors, and that's uncomfortably close to the freq-weight-into-target path already in the tree (`sp[mask] = species_freq_weights[mask]`), which collapsed bats because it turns the binary target soft and the model then learns to suppress the species. So I'd be careful extending propagation that way without strong range priors.

I don't see these as competing, though. Propagation fixes spatial under-sampling; the loss `weight` handles the residual gradient-share imbalance that survives even with perfect coverage, since GBIF will always hold far fewer bat records. Even ideal propagation leaves a big gap there.

If the per-class scalar feels too blunt (it is a bit, `mammalia=6` hits all 879 mammals including the well-recorded ones), I'm happy to swap it for a frequency-derived per-species weight: continuous inverse-frequency from observed cell-weeks, clipped. That's recalibration, still just BCE's `weight=`, with no manual knob per run. And I'm glad to prototype a range-aware extension to propagation and smoothing and A/B all three (propagation only, weight only, both) on the same recipe, so GeoScore and the geographic contrast decide it rather than us arguing it in the abstract.

**Comment by kahst** [comment]
Ah, yes, thanks. If you have the means to try all three variants, we can decide which one we like better. I know, label propagation is cluttered and can probably be a little more concise, but ideally we can tag it on there. But I'm open minded, as long as the overall geo score doesn't suffer and bat geo score goes up, we should be fine.
